File: pset6/readability/readability.py
def main():
    # Recieve the text to be graded as input from user
    text = input("Enter the text to be graded: ")

    # Find the number of sentences, words and letters in the user input
    lines, words, letters = count_str(text)

    # Find out the grade of the text using  the Coleman-Liau index
    grade = coleman_liau(lines, words, letters)

    # Print grade
    print_grade(grade)


def count_str(text):
    """Function to keep count of the number of sentences, words and letters in a given text"""
    lines, words, letters = 0, 1, 0

    # Iterate through the text and keep count of the number of sentences, words and letters
    for char in text:
        if char.isalpha():
            letters += 1
        elif char.isspace():
            words += 1
        elif char in [".", "?", "!"]:
            lines += 1

    # Return the number of sentences, words and letters
    return lines, words, letters


def coleman_liau(lines, words, letters):
    # Calculate the Coleman-Liau index for the given text
    return 0.0588 * ((letters / words) * 100) - 0.296 * ((lines / words) * 100) - 15.8


def print_grade(grade):
    """Function to print the approprite grade for the text according to the Coleman-Liau index"""
    if grade < 1:
        print("Before Grade 1")
    elif grade > 16:
        print("Grade 16+")
    else:
        print(f"Grade {round(grade)}")


main()

File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int get_letters(string text);

int get_words(string text);

int get_sentences(string text);

int main(void)
{
    // Accept a text as input
    string text = get_string("Text: ");
    
    // Get number of letters, words and sentences in the text
    int letters = get_letters(text);
    int words = get_words(text);
    int sentences = get_sentences(text);
    
    // L --> number of letters per 100 words
    float L = (letters / (float) words) * 100;
    // S --> number of words per 100 sentences
    float S = (sentences / (float) words) * 100;
    
    // Calculating the grade of the text using the Coleman Liau index
    int coleman_liau_index = round((0.0588 * L) - (0.296 * S) - 15.8);
    
    // Printing the text's grade
    if (coleman_liau_index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (coleman_liau_index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", coleman_liau_index);
    }
}

int get_letters(string text)
{
    // Letters keeps count of number of letters in the text
    int letters = 0;
    // Iterating through each character in the text
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        char lowercase_char = tolower(text[i]);
        // checking whether a character is a letter and if true, adding 1 to letters
        if (lowercase_char >= 97 && lowercase_char <= 122)
        {
            letters++;
        }
    }
    return letters;
}

int get_words(string text)
{
    int words = 1;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            words ++;
        }
    }
    return words;
}

int get_sentences(string text)
{
    int sentences = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '.' || text[i] == '?' || text[i] == '!')
        {
            sentences++;
        }
    }
    return sentences;
}